diff --git a/pngcrush.c b/pngcrush.c
index ea688b9..1cbdaed 100644
--- a/pngcrush.c
+++ b/pngcrush.c
@@ -1699,7 +1699,6 @@ static int overwrite = 0; /* overwrite the input file instead of creating
                            a new output file */
 static int nofilecheck = 0;
 static int no_limits = 0;
-static int new_mng = 0;
 static png_bytep row_buf;
 #ifdef PNGCRUSH_MULTIPLE_ROWS
 static png_bytepp row_pointers;
@@ -2627,7 +2626,6 @@ void pngcrush_write_png(png_structp write_pointer, png_bytep data,
 static void pngcrush_flush(png_structp png_ptr)
 {
    /* Do nothing. */
-   PNGCRUSH_UNUSED(png_ptr)
 }
 
 
@@ -2656,6 +2655,7 @@ void pngcrush_examine_pixels_fn(png_structp png_ptr, png_row_infop
       if ((row_info->color_type == 2 || row_info->color_type == 6) &&
           make_gray == 1) /* RGB */
       {
+        make_gray = 2;
         if (row_info->bit_depth == 8)
           {
             int incr=3;
@@ -2741,6 +2741,7 @@ void pngcrush_examine_pixels_fn(png_structp png_ptr, png_row_infop
                make_opaque == 1))
       {
         i=(int) row_info->rowbytes-1;
+        make_gray = 2;

         if (row_info->bit_depth == 8)
           {
@@ -7190,7 +7191,7 @@ png_uint_32 pngcrush_measure_idat(png_structp png_ptr)
         png_reset_crc(png_ptr);
         png_crc_read(png_ptr, chunk_name, 4);
 
-        if (new_mng)
+        if (0)
         {
           const png_byte png_DHDR[5] = {  68, 72, 68, 82, '\0' };
           const png_byte png_DEFI[5] = {  68, 69, 70, 73, '\0' };
@@ -7357,7 +7357,7 @@ png_uint_32 pngcrush_measure_idat(png_structp png_ptr)
         }
 
 #ifdef PNG_UINT_acTL
-        else if (png_get_uint_32(chunk_name) == PNG_UINT_acTL)
+         if (png_get_uint_32(chunk_name) == PNG_UINT_acTL)
 #else
         else if (!png_memcmp(chunk_name, png_acTL, 4))
 #endif
@@ -7566,7 +7566,7 @@ png_uint_32 pngcrush_measure_idat(png_structp png_ptr)
                 { 77, 69, 78, 68, '\0' };
             if (!png_memcmp(chunk_name, png_MEND, 4))
             {
-                if (new_mng)
+                if (0)
                 {
                     png_free(mng_ptr,bb);
                     return (0);
@@ -7638,6 +7638,7 @@ void print_version_info(void)
        * immediately after this sentence: */
       " |    Copyright (C) 1998-2002, 2006-2013 Glenn Randers-Pehrson\n"
       " |    Portions copyright (C) 2005       Greg Roelofs\n"
+      " | Modified version for ImageOptim.\n"
       " | This is a free, open-source program.  Permission is irrevocably\n"
       " | granted to everyone to use this version of pngcrush without\n"
       " | payment of any fee.\n"
