diff --git a/pngcrush.c b/pngcrush.c
--- a/pngcrush.c
+++ b/pngcrush.c
@@ -1513,7 +1513,7 @@
 #endif
 
 #ifndef LIBPNG_UNIFIED
-#include <png.h>
+#include "png.h"
 #define PNGCRUSH_TIMER_UINT_API extern unsigned int PNGAPI
 #define PNGCRUSH_TIMER_VOID_API extern void PNGAPI
 #else
@@ -2295,7 +2295,6 @@
                                 creating a new output file */
 static int nofilecheck = 0;
 static int no_limits = 0;
-static int new_mng = 0;
 static png_bytep row_buf;
 #ifdef PNGCRUSH_MULTIPLE_ROWS
 static png_bytepp row_pointers;
@@ -2845,7 +2844,6 @@
         char keystroke;
         fprintf(STDERR, "Press [ENTER] key to continue.\n");
         keystroke = (char) getc(stdin);
-        keystroke = keystroke;  /* stifle compiler warning */
     }
 }
 
@@ -3234,7 +3232,6 @@
 static void pngcrush_flush(png_structp png_ptr)
 {
    /* Do nothing. */
-   PNGCRUSH_UNUSED(png_ptr)
 }
 
 
@@ -3287,6 +3284,7 @@
       if ((row_info->color_type == 2 || row_info->color_type == 6) &&
           make_gray == 1) /* RGB */
       {
+        make_gray = 2;
         if (row_info->bit_depth == 8)
           {
             int incr=3;
@@ -3372,6 +3370,7 @@
                make_opaque == 1))
       {
         i = (int) row_info->rowbytes-1;
+        make_gray = 2;
 
         if (row_info->bit_depth == 8)
           {
@@ -4084,7 +4083,6 @@
             names++;
             BUMP_I;
             mngname = argv[i];
-            new_mng++;
         }
 #endif
 
@@ -4828,7 +4826,7 @@
             number_of_open_files++;
 
 #ifdef PNGCRUSH_LOCO
-            if (new_mng)
+            if (0)
             {
 
 #  ifdef PNG_USER_MEM_SUPPORTED
@@ -4866,7 +4864,7 @@
             idat_length[0] = measure_idats(fpin);
 
 #ifdef PNGCRUSH_LOCO
-            if (new_mng)
+            if (0)
             {
                     png_destroy_write_struct(&mng_ptr, NULL);
                     FCLOSE(mng_out);
@@ -8243,7 +8241,7 @@
                 (unsigned long) pngcrush_get_uint_31(png_ptr,&buff[24]));
             }
 
-            if (new_mng)
+            if (0)
             {
                 /* write the MNG 8-byte signature */
                 pngcrush_write_png(mng_ptr, &mng_signature[0],
@@ -8343,7 +8341,7 @@
               printf("   Reading %c%c%c%c chunk.\n",
                   chunk_name[0],chunk_name[1],chunk_name[2],chunk_name[3]);
 
-        if (new_mng)
+        if (0)
         {
           const png_byte png_DHDR[5] = {  68, 72, 68, 82, '\0' };
           const png_byte png_DEFI[5] = {  68, 69, 70, 73, '\0' };
@@ -8653,7 +8651,7 @@
                 { 77, 69, 78, 68, '\0' };
             if (!png_memcmp(chunk_name, png_MEND, 4))
             {
-                if (new_mng)
+                if (0)
                 {
                     png_free(mng_ptr,bb);
                     return (0);
@@ -8736,6 +8734,7 @@
        * immediately after this sentence: */
       " |    Copyright (C) 1998-2002, 2006-2017 Glenn Randers-Pehrson\n"
       " |    Portions Copyright (C) 2005 Greg Roelofs\n"
+      " | Modified version for ImageOptim.\n"
       " | This is a free, open-source program.  Permission is irrevocably\n"
       " | granted to everyone to use this version of pngcrush without\n"
       " | payment of any fee.\n"
