diff --git a/src/opngoptim.c b/src/opngoptim.c
index aab5580..ceaa60f 100644
--- a/src/opngoptim.c
+++ b/src/opngoptim.c
@@ -1091,6 +1091,8 @@ opng_read_file(FILE *infile)
          reductions &= ~OPNG_REDUCE_BIT_DEPTH;
       if (options.nc)
          reductions &= ~OPNG_REDUCE_COLOR_TYPE;
+      if (options.nd)
+         reductions &= ~OPNG_REDUCE_DIRTY_ALPHA;
       if (options.np)
          reductions &= ~OPNG_REDUCE_PALETTE_ALL;
       if (options.nz && (process.status & INPUT_HAS_PNG_DATASTREAM))
diff --git a/src/opngreduc.c b/src/opngreduc.c
index 1f01401..ac37521 100644
--- a/src/opngreduc.c
+++ b/src/opngreduc.c
@@ -515,7 +515,7 @@ opng_reduce_bits(png_structp png_ptr, png_infop info_ptr,
       src_bit_depth, dest_bit_depth, src_byte_depth, dest_byte_depth,
       src_color_type, dest_color_type, src_channels, dest_channels,
       src_sample_size, dest_sample_size, src_offset_alpha;
-   unsigned int tran_tbl[8], k;
+   unsigned int tran_tbl[8]={0}, k;
 
    png_debug(1, "in opng_reduce_bits\n");
 
@@ -1105,6 +1105,69 @@ opng_analyze_sample_usage(png_structp png_ptr, png_infop info_ptr,
       usage_map[info_ptr->background.index] = 1;
 }
 
+/*
+ * Remove RGB components or transparent pixels in RGB+alpha images.
+ * The function returns OPNG_REDUCE_DIRTY_ALPHA if any pixels were cleared.
+ */
+png_uint_32 /* PRIVATE */
+opng_reduce_dirty_alpha(png_structp png_ptr, png_infop info_ptr)
+{
+   png_uint_32 result = OPNG_REDUCE_NONE;
+   png_bytepp row_ptr;
+   png_bytep sample_ptr, alpha_row;
+   png_uint_32 height, width, channels, i, j;
+   unsigned int color_type;
+
+   png_debug(1, "in opng_reduce_dirty_transparency\n");
+
+   if (info_ptr->bit_depth != 8)
+      return OPNG_REDUCE_NONE;  /* nothing is done in this case */
+
+   color_type = info_ptr->color_type;
+   OPNG_ASSERT(!(info_ptr->color_type & PNG_COLOR_MASK_PALETTE));
+
+   row_ptr   = info_ptr->row_pointers;
+   height    = info_ptr->height;
+   width     = info_ptr->width;
+   channels  = info_ptr->channels;
+   alpha_row = (png_bytep)png_malloc(png_ptr, width);
+
+   /* Search for transparent pixels. */
+   for (i = 0; i < height; ++i, ++row_ptr)
+   {
+      sample_ptr = *row_ptr;
+      opng_get_alpha_row(png_ptr, info_ptr, *row_ptr, alpha_row);
+      if (color_type & PNG_COLOR_MASK_COLOR)
+      {
+         for (j = 0; j < width; ++j, sample_ptr += channels)
+         {
+            if (alpha_row[j] == 0)
+            {
+               sample_ptr[0] = 0;
+               sample_ptr[1] = 0;
+               sample_ptr[2] = 0;
+               result = OPNG_REDUCE_DIRTY_ALPHA;
+            }
+         }
+      }
+      else  /* grayscale */
+      {
+         for (j = 0; j < width; ++j, sample_ptr += channels)
+         {
+            if (alpha_row[j] == 0)
+            {
+               sample_ptr[0] = 0;
+               result = OPNG_REDUCE_DIRTY_ALPHA;
+            }
+         }
+      }
+   }
+
+   png_free(png_ptr, alpha_row);
+   return result;
+}
+
+
 
 /*
  * Reduce the palette (only the fast method is implemented).
@@ -1292,7 +1355,7 @@ opng_reduce_image(png_structp png_ptr, png_infop info_ptr,
    png_uint_32 reductions)
 {
    unsigned int color_type;
-   png_uint_32 result;
+   png_uint_32 result = 0;
 
    png_debug(1, "in opng_reduce_image_type\n");
 
@@ -1317,8 +1380,14 @@ opng_reduce_image(png_structp png_ptr, png_infop info_ptr,
 
    /* The reductions below must be applied in the given order. */
 
+   if ((color_type == PNG_COLOR_TYPE_RGB_ALPHA ||
+        color_type == PNG_COLOR_TYPE_GRAY_ALPHA)
+       && (reductions & OPNG_REDUCE_DIRTY_ALPHA)
+      )
+      result |= opng_reduce_dirty_alpha(png_ptr, info_ptr);
+
    /* Try to reduce the high bits and color/alpha channels. */
-   result = opng_reduce_bits(png_ptr, info_ptr, reductions);
+   result |= opng_reduce_bits(png_ptr, info_ptr, reductions);
 
    /* Try to reduce the palette image. */
    if (color_type == PNG_COLOR_TYPE_PALETTE &&
diff --git a/src/opngreduc.h b/src/opngreduc.h
index 1877058..e5d6443 100644
--- a/src/opngreduc.h
+++ b/src/opngreduc.h
@@ -83,6 +83,8 @@ extern PNG_EXPORT(png_uint_32,opng_reduce_image)
                                                 from PLTE and tRNS */
 #define OPNG_REDUCE_PALETTE_FAST     0x0200  /* remove trailing sterile entries
                                                 from PLTE and tRNS */
+#define OPNG_REDUCE_DIRTY_ALPHA      0x0400  /* zero RGB components of
+                                                transparent pixels */
 #define OPNG_REDUCE_ANCILLARY        0x1000
 
 #define OPNG_REDUCE_BIT_DEPTH  \
@@ -98,6 +100,6 @@ extern PNG_EXPORT(png_uint_32,opng_reduce_image)
 
 #define OPNG_REDUCE_ALL  \
    (OPNG_REDUCE_BIT_DEPTH | OPNG_REDUCE_COLOR_TYPE | \
-    OPNG_REDUCE_PALETTE_ALL | OPNG_REDUCE_ANCILLARY)
+    OPNG_REDUCE_PALETTE_ALL | OPNG_REDUCE_ANCILLARY | OPNG_REDUCE_DIRTY_ALPHA)
 
 #endif /* OPNG_IMAGE_REDUCTIONS_SUPPORTED */
diff --git a/src/optipng.c b/src/optipng.c
index 5fc1ca1..6f9c710 100644
--- a/src/optipng.c
+++ b/src/optipng.c
@@ -109,6 +109,7 @@ static const char *msg_help =
    "    -nb\t\t\tno bit depth reduction\n"
    "    -nc\t\t\tno color type reduction\n"
    "    -np\t\t\tno palette reduction\n"
+   "    -nd\t\t\tno dirty alpha transparency cleaning\n"
 #if 0  /* not implemented */
    "    -nm\t\t\tno metadata optimization\n"
 #endif
@@ -379,6 +380,10 @@ parse_args(int argc, char *argv[])
             options.nb = options.nc = options.np = 1;
             /* options.nm = 1; */
         }
+        else if (strcmp(opt, "nd") == 0)
+        {
+            options.nd = 1;
+        }
         else if (strcmp("nz", opt) == 0)
         {
             options.nz = 1;
@@ -668,9 +673,9 @@ app_print_cntrl(int cntrl_code)
     }
 
     if (con_file != NULL)
-        fputs(con_str, con_file);
+        fprintf(con_file, "%s",con_str);
     if (log_file != NULL)
-        fputs(log_str, log_file);
+        fprintf(log_file, "%s",log_str);
 }
 
 
diff --git a/src/optipng.h b/src/optipng.h
index 646ccaa..f671b21 100644
--- a/src/optipng.h
+++ b/src/optipng.h
@@ -31,7 +31,7 @@ struct opng_options
     int full;
     int interlace;
     int keep;
-    int nb, nc, np, nz;
+    int nb, nc, np, nz, nd;
     int preserve;
     int quiet;
     int simulate;
diff --git a/src/proginfo.h b/src/proginfo.h
index eac9b1c..11540f3 100644
--- a/src/proginfo.h
+++ b/src/proginfo.h
@@ -1,5 +1,5 @@
 #define PROGRAM_NAME        "OptiPNG"
 #define PROGRAM_DESCRIPTION "Advanced PNG optimizer"
-#define PROGRAM_VERSION     "0.6.4"
+#define PROGRAM_VERSION     "0.6.4 + dirtyalpha patch"
 #define PROGRAM_COPYRIGHT   "Copyright (C) 2001-2010 Cosmin Truta"
 #define PROGRAM_URI         "http://optipng.sourceforge.net/"
