diff --git a/lib/pngxtern/pngxread.c b/lib/pngxtern/pngxread.c
index e0744d3..a3fc2db 100644
--- a/lib/pngxtern/pngxread.c
+++ b/lib/pngxtern/pngxread.c
@@ -79,10 +79,8 @@ pngx_read_image(png_structp png_ptr, png_infop info_ptr,
 {
    png_byte sig[128];
    size_t num;
-   int (*read_fn)(png_structp, png_infop, FILE *);
    FILE *stream;
    fpos_t fpos;
-   int result;
 
    /* Precondition. */
 #ifdef PNG_FLAG_MALLOC_NULL_MEM_OK
@@ -110,26 +108,5 @@ pngx_read_image(png_structp png_ptr, png_infop info_ptr,
       }
       return 1;
    }
-
-   /* Check the signature bytes against other known image formats. */
-   if (pngx_sig_is_bmp(sig, num, fmt_name, fmt_description) > 0)
-      read_fn = pngx_read_bmp;
-   else if (pngx_sig_is_gif(sig, num, fmt_name, fmt_description) > 0)
-      read_fn = pngx_read_gif;
-   else if (pngx_sig_is_jpeg(sig, num, fmt_name, fmt_description) > 0)
-      read_fn = pngx_read_jpeg;
-   else if (pngx_sig_is_pnm(sig, num, fmt_name, fmt_description) > 0)
-      read_fn = pngx_read_pnm;
-   else if (pngx_sig_is_tiff(sig, num, fmt_name, fmt_description) > 0)
-      read_fn = pngx_read_tiff;
-   else
-      return 0;  /* not a known image format */
-
-   /* Read the image. */
-   result = read_fn(png_ptr, info_ptr, stream);
-   /* Signature checking may give false positives; reading can still fail. */
-   if (result <= 0)  /* this isn't the format we thought it was */
-      if (fsetpos(stream, &fpos) != 0)
-         png_error(png_ptr, "Can't fseek in input file stream");
-   return result;
+   return 0;
 }
